tools/autograd/templates/python_variable_methods.cpp

```diff
@@ -279,34 +280,63 @@ static PyObject * THPVariable_invert(PyObject* self, PyObject* args) {
   END_HANDLE_TH_ERRORS
 }
 
-static Tensor dispatch_to_backend(const Tensor & self, Backend backend) {
+static void lazy_init_cuda() {
+  static std::once_flag once;
+  std::call_once(once, []() {
+    auto module = THPObjectPtr(PyImport_ImportModule("torch.cuda"));
+    if (!module) throw python_error();
+    auto res = THPObjectPtr(PyObject_CallMethod(module.get(), "_lazy_init", ""));
+    if (!res) throw python_error();
+  });
+}
+
+static Tensor dispatch_type(const Tensor & self, const at::Type & type, int device=-1, bool async=false) {
+  if (type.is_cuda()) {
+    lazy_init_cuda();
+  }
   AutoNoGIL no_gil;
-  AutoGPU auto_gpu(self);
-  return self.toBackend(backend);
+  AutoGPU auto_gpu(device != -1 ? device : self.type().is_cuda() ? self.get_device() : -1);
+  if (self.type() == type && self.type().is_cuda() && self.get_device() != at::current_device()) {
+    return type.copy(self, async);
+  }
+  if (!self.type().is_cuda() && type.is_cuda()) {
+    return type.copy(self, async);
+  }
+  return self.toType(type);
```

Comment:
why can't we just type.copy here as well?

Reply:
We don't want to force a copy if the types and the devices are the same.

Reply:
Sorry, I expressed this poorly.  I meant what you are doing here seems totally natural and should have better ATen support.  Instead you have to go through all these contortions to figure out if you can avoid a copy.

If at::current_device returns -1 then there is no reason that Tensor.get_device can't, which means at a minimum you can collapse the first two ifs by just looking at the current devices and types (you can do this now by having a "safe_device = self.is_cuda() ? self.get_device : -1" or something.

Reply:
I've combined the if statements.

1) I like the idea of making `Tensor.get_device()` return -1 for CPU tensors. I think it'll simplify the calling code in a bunch of places. I'll do that in a subsequent PR
2) It would be nice to push all of this logic down to ATen. The problem is that ATen doesn't have the "auto gpu" semantics (because it might affect perf), so it doesn't translate as well.

Reply:
Sounds good.